File: src/main/java/project/interactivenovelplatform/service/ChatService.java
package project.interactivenovelplatform.service;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Slice;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.multipart.MultipartFile;
import project.interactivenovelplatform.dto.request.ChatSettingsRequestDto;
import project.interactivenovelplatform.dto.request.CreateGroupRequest;
import project.interactivenovelplatform.dto.request.SendMessageRequestDto;
import project.interactivenovelplatform.dto.response.ConversationResponseDto;
import project.interactivenovelplatform.dto.response.MessageResponseDto;

import java.util.List;

public interface ChatService {
    ConversationResponseDto getOrCreatePrivateChat(Long currentUserId, Long targetUserId);
    ConversationResponseDto createGroupChat(Long creatorId, CreateGroupRequest request);
    void sendMessage(Long senderId, SendMessageRequestDto dto, List<MultipartFile> files);
    Page<ConversationResponseDto> getUserChats(Long userId, Pageable pageable);
    Slice<MessageResponseDto> getChatMessages(Long userId, Long conversationId, Pageable pageable);
    void sendTypingStatus(Long userId, Long conversationId);
    void deleteMessage(Long requesterId, Long messageId);
    void deleteChatForUser(Long userId, Long conversationId);
    void leaveGroup(Long userId, Long conversationId);
    void toggleChatSettings(Long userId, ChatSettingsRequestDto settings);
    void kickUser(Long adminId, Long targetUserId, Long conversationId);

    ConversationResponseDto addUserToGroup(Long inviterId, List<Long> targetUserIds, Long conversationId);

    @Transactional
    void markConversationAsRead(Long userId, Long conversationId);
}

File: frontend/src/api/chatService.ts
import apiClient from "@/api/axios.ts";
import type {
  ChatSettingsRequestDto,
  ConversationResponseDto,
  CreateGroupRequest,
  MessageResponseDto,
  SendMessageRequestDto
} from "@/types/chat.ts";
import type {PagedModel, SliceModel} from "@/types/PagedModel.ts";


export async function getMyChats(page: number = 0, size: number = 20 ): Promise<PagedModel<ConversationResponseDto>> {
  const response = await apiClient.get('/chats', { params: { page, size} });
  return response.data;
}


export async function getChatMessages(conversationId: number, page: number = 0, size: number = 50,sort: string = 'timestamp,desc'): Promise<SliceModel<MessageResponseDto>> {
  const response = await apiClient.get(`/chats/${conversationId}/messages`, { params: { page, size,sort } });
  return response.data;
}


export async function getOrCreatePrivateChat(targetUserId: number): Promise<ConversationResponseDto> {
  const response = await apiClient.post(`/chats/private/${targetUserId}`);
  return response.data;
}


export const createGroupChat = async (payload: CreateGroupRequest): Promise<ConversationResponseDto> => {
  const formData = new FormData();

  formData.append('title', payload.title);

  payload.memberIds.forEach(id => {
    formData.append('memberIds', id.toString());
  });

  if (payload.avatarUrl) {
    formData.append('avatarUrl', payload.avatarUrl);
  }
  const response = await apiClient.post<ConversationResponseDto>('/chats/group', formData, {
    headers: {
      'Content-Type': 'multipart/form-data'
    }
  });
  return response.data;
}


export async function sendMessage(conversationId: number, dto: SendMessageRequestDto, files?: File[]): Promise<void> {
  const formData = new FormData();

  
  formData.append('dto', new Blob([JSON.stringify(dto)], { type: 'application/json' }));

  
  if (files && files.length > 0) {
    files.forEach(file => formData.append('files', file));
  }

  const response = await apiClient.post(`/chats/${conversationId}/messages`, formData, {
    headers: {
      'Content-Type': 'multipart/form-data',
    },
  });
  return response.data;
}


export async function deleteMessage(messageId: number): Promise<void> {
  const response = await apiClient.delete(`/chats/messages/${messageId}`);
  return response.data;
}


export async function deleteChatForUser(conversationId: number): Promise<void> {
  const response = await apiClient.delete(`/chats/${conversationId}`);
  return response.data;
}


export async function toggleChatSettings(settings: ChatSettingsRequestDto): Promise<void> {
  const response = await apiClient.patch(`/chats/conversation/settings`, settings);
  return response.data;
}


export async function leaveGroup(conversationId: number): Promise<void> {
  const response = await apiClient.post(`/chats/${conversationId}/leave`);
  return response.data;
}


export async function addUserToGroup(conversationId: number, targetUserIds: Array<number>): Promise<ConversationResponseDto> {
  const response = await apiClient.post(`/chats/${conversationId}/members`, targetUserIds);
  return response.data;
}


export async function kickUser(conversationId: number, targetUserId: number): Promise<void> {
  const response = await apiClient.delete(`/chats/${conversationId}/members/${targetUserId}`);
  return response.data;
}


export async function sendTypingStatus(conversationId: number): Promise<void> {
  const response = await apiClient.post(`/chats/${conversationId}/messages/typing`);
  return response.data;
}

export async function markConversationAsRead(conversationId:number): Promise<void> {
  const response = await apiClient.post(`/chats/${conversationId}/read`);
  return response.data;
}
